Ames - packets, diagrams, serf and king

syzygyzer opened the pull request:
This is the second of several Ames PR's I've been working on (the first being #943), which are mostly being done now for the sake of the security audit but of course the information is widely useful. It is mostly an elaboration on what was already summarized at the top.

This adds two major sections:
- Packets: what format they come in, how they are encrypted, how they are serialized, how messages are packetized, how they are acked, and how they are relayed.
- The Serf and the King: The Serf does most of the grunt work for Ames, but the King has important responsibilities too.

Each section is accompanied by a diagram that helps to describe their content.

There are still some important things to come. A discussion of nacks and failures in Ames in general, how Ames and Jael work together to handle keys, and what specifically constitutes an Arvo event and how that relates to flows. If there are other parts of Ames you'd like documented, now is the time to mention it.

syzygyzer commented:
I've added content on flows, nacks, and naxplanations, and reworded a few other things as appropriate.

syzygyzer commented:
pinging @belisarius222

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+
+#### (N)ack packets
+
+The contents of a given (n)ack packet are deterministic. The `content` (i.e. the
```
Maybe a more precise way of stating what you mean here would be to say something like:
```
A fragment ack's contents (before encryption) are a pure function of the following data:
[our-life her-life ...]

This means all re-sends of an ack packet will be bitwise identical to each other, unless one of the peers changes its encryption keys.
```

The same is true for message acks and message fragment packets, too.  The unifying concept here, which I'm not sure is really explained anywhere, is that before encryption (which introduces a dependency on the sender's and receiver's life numbers) and relaying (which can modify the `origin` field), every Ames packet is a pure function of `[sndr rcvr flow-number message-number packet-id]`, where `packet-id` varies by packet type, but is either `[%message-fragment fragment-number]`, `[%fragment-ack fragment-number]`, or `[%message-ack nack=(unit naxplanation)]`.

syzygyzer replied:
https://github.com/urbit/docs/pull/957/commits/b9353544c7d1840c7ec464e916efb088462c0e12#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR234-R240

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+the ack-receiving ship's networking keys. `bone` is an opaque number identifying the flow, `message-num` denotes the number of the
+message in the flow identified by `bone`, and
+`fragment-num` denotes the number of the fragment of the message identified by
+`message-num` being acked. 
```
The "being acked" at the end of the sentence is hard to parse.  Maybe split this into multiple sentences?

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+`fragment-num` denotes the number of the fragment of the message identified by
+`message-num` being acked. 
+
+A message (n)ack is an extended fragment ack that is obtained by encrypting the
```
> extended fragment ack
Not really.  It's a different kind of ack, with different contents.

belisarius222 commented on tutorials/arvo/ames.md:
````diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+`+jam` of the following noun:
+
+```hoon
+[our-life her-life bone message-num fragment-num ok=? lag=@dr]
````
There's no `fragment-num` in a message ack.  The ack packet attests to having processed the whole message, so it doesn't need to refer to any particular fragment.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+the packets which make them up need not be. `%boon`s are totally ordered in the
+same fashion, but there is not a coordinated ordering between `%plea`s and
+`%boon`s in a given flow beyond the implicit one arising from the fact that a
+`%boon` is always a response to a `%plea`.
```
I think it would be good to clarify here that each `%plea` can give rise to zero, one, or many `%boon`s.

belisarius222 replied:
You could also say here that to create a new flow, a ship sends a `%plea` to another ship.  Only the creator of a flow can send a `%plea` on it, and only the other party can send a `%boon` on it.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+referred to by its first bone. The set of types of packets is partitioned into
+four subsets, each of which is assigned a bone. Each bone is a one-way street
+for packets to travel along, so e.g. acks to packets making up a `%plea` are
+sent along different bones.
```
They're not sent on different bones from each other; all acks to `%plea`s on a given flow will be on the same bone.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+Inside of Ames, each flow has four sequential opaque `@ud`s
+called bones that are unique to that flow. Thus the flow itself is often
+referred to by its first bone. The set of types of packets is partitioned into
+four subsets, each of which is assigned a bone. Each bone is a one-way street
```
> The set of types of packets is partitioned into four subsets, each of which is assigned a bone.

I find this sentence hard to understand, and this concept is explained in more detail later.  I'd remove the sentence.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+ - bone 15, naxplanations to `~bacbel-tagfeg`.
+
+Each bone is handled separately by congestion control, and this is the reason
+for their segregation. For example, say a two packet `%plea` is sent on bone 12,
```
"two packet" -> "two-packet"

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+ - bone 14, acks of naxplanations to `~worwel-sipnum`,
+ - bone 15, naxplanations to `~bacbel-tagfeg`.
+
+Each bone is handled separately by congestion control, and this is the reason
```
Separate congestion control is one reason to separate bones.  There are two others.

It's important that I be able to start a flow with you without risk of race conditions.  That is, if I try to start a flow with you at the same time you try to start a flow with me, when we receive each other's packets, that shouldn't cause a conflict.  Flipping the last bit of the bone based on whether a packet is an ack or a message fragment allows for me to create flow 8 with you without that ever coming into conflict with the flow 8 that you created with me.

The reason to have four bones per flow, not just two, is that naxplanation packets shouldn't be able to conflict with happy-path packets.  Since naxplanations are generated dynamically in response to hearing messages, if they were  on the same bones as happy-path messages, they would have to occupy new numbers within those flows.  Maybe it would be possible to get that to work, but it was easier to be confident in the correctness of the design with naxplanations on their own bones that can't conflict with the happy path.

syzygyzer replied:
https://github.com/urbit/docs/pull/957/commits/d42f0a3aee11fd2b410448332da010a1a880ac23#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR302-R320

How's this?

syzygyzer replied:
Whoops, somehow started a new review for that comment. I'm gonna leave it just to ensure I don't fuck something else up.

belisarius222 replied:
LGTM

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+ - The payload of this packet will be the `+jam` of `[origin content]`. `origin`
+for this initial packet will be `~`, which implies that the packet is
+originating from the sender. `content` is the `+jam` of the message, encrypted
+using a Diffie-Hellman symmetric key obtained with the private key of
```
I think this phrasing is a bit ambiguous.  The symmetric key is obtained using a Diffie-Hellman, once, *and then* the encryption of each packet is performed with that symmetric key.

syzygyzer replied:
https://github.com/urbit/docs/pull/957/commits/3f725698e167a6d4ff1cf726afbe515cb322a755#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daL351-R366

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+Once `~worwel-sipnum` processes the packet, it will know the IP and port of
+`~bacbel-tagfeb` since `~zod` included it when it forwarded the packet and so
+will send an ack packet directly to `~bacbel-tagfeb`. Communication between the
+two ships will now be direct until one of them changes their IP address, port, or networking keys.
```
NATs and firewalls can prevent a direct peer-to-peer connection, even if the two ships know each other's IP and port.  In that case, a star or galaxy will continue to relay packets.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+## The Serf and the King
+
+Urbit's functionality is split between the two binaries `urbit-worker` (sometimes
+called the Serf) and `urbit-king` (sometimes called the King). This division of
```
I think you mean `urbit-daemon`, not `urbit-king`.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+
+In short, the Serf is the Nock runtime and so keeps track of the current state
+of Arvo as a Nock noun and updates the state by `%poke`ing it with nouns, and
+then informs the King of the new state. The King manages snapshots of the Arvo
```
It's actually the worker, not the daemon, that manages Arvo snapshots, since that's the process that holds the Arvo state in memory.

syzygyzer replied:
I'm a bit confused - I thought the daemon and the worker were the same thing? Maybe I should be more clear of what is meant by manage - does the King save the snapshots to disk since it handles I/O? Or does the daemon (or worker? whichever you mean) actually also talk to Unix in order to save snapshots?

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+
+The King has several submodules, one of them being an Ames I/O submodule. This
+submodule is responsible for wrapping outgoing Ames packets as UDP packets, and
+looking at incoming UDP packets handed to it by Unix to ensure that they are
```
It doesn't do much validation; it mostly turns the packets into Ames packets and sends them to the worker to be processed.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+The King has several submodules, one of them being an Ames I/O submodule. This
+submodule is responsible for wrapping outgoing Ames packets as UDP packets, and
+looking at incoming UDP packets handed to it by Unix to ensure that they are
+valid Ames packets. It also maintains an incoming and outgoing packet queue.
```
Outgoing packets are essentially fired off immediately, not really enqueued.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+
+"Unix", in this case, is actually the King. The King receives the `%send`
+instruction, wraps the packet contained within as a UDP packet, and queues the
+UDP packet contained. Once it is time to send it (which will be determined by
```
It's the Ames vane that performs congestion control logic, not the daemon process.  Once the daemon process is informed it's supposed to send a UDP packet, it fires it off without question.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+networking interface, at which point it is no longer controlled by any part of Urbit.
+
+Now the receiving King is handed a UDP packet by Unix. The King removes the UDP
+wrapped and checks to see that it has received a valid Ames packet. If so, it
```
"wrapped" -> "wrapper"

belisarius222 replied:
The daemon also doesn't fully check packet validity before sending it to be processed by the Ames vane.

belisarius222 commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+
+Now the receiving King is handed a UDP packet by Unix. The King removes the UDP
+wrapped and checks to see that it has received a valid Ames packet. If so, it
+`+cue`s the payload and wraps it as a `%hear` `note` and passes it to the Serf.
```
The daemon doesn't `+cue` the payload; it jams the lane on which it heard the packet, but other than that, it delivers the packet to the Ames vane as an atom by coping the bytes it heard on its UDP port.

syzygyzer commented:
@belisarius222 I plan to finish addressing your review today or tomorrow. I am wondering though, with the upcoming vere release, does the serialization steps described here change? I discuss two serialization steps, but I remember during the infra meeting last week (I think) it was mentioned that there were even more than that?

"- Improved event loop performance by removing unnecessary serialization."

belisarius222 commented:
That's an unrelated issue, I believe. We might change the packet format in other ways when we breach, but that isn't set in stone yet.

—
~rovnys-ricfer
https://urbit.org

On Mon, Nov 16 2020 at 11:29 AM, poprox < notifications@github.com > wrote:

> 
> 
> 
> 
> 
> 
> 
> @belisarius222 ( https://github.com/belisarius222 ) I plan to finish
> addressing your review today or tomorrow. I am wondering though, with the
> upcoming vere release, does the serialization steps described here change?
> I discuss two serialization steps, but I remember during the infra meeting
> last week (I think) it was mentioned that there were even more than that?
> 
> 
> 
> "- Improved event loop performance by removing unnecessary serialization."
> 
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub (
> https://github.com/urbit/docs/pull/957#issuecomment-728172853 ) , or unsubscribe
> (
> https://github.com/notifications/unsubscribe-auth/AAGVR5OKMOBDXWPGBEQZTX3SQFHP7ANCNFSM4SGSUOJQ
> ).
> 
> 
>

syzygyzer commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+`%boon`s and naxplanations are never nacked.
+Individual packets are never nacked, only complete `%plea` messages. This works
+because only complete packets ever make it to Ames, as malformed
+packets will be discarded by the Ames I/O module in the
```
https://github.com/urbit/docs/pull/957/commits/b9353544c7d1840c7ec464e916efb088462c0e12#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR216

syzygyzer commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+because only complete packets ever make it to Ames, as malformed
+packets will be discarded by the Ames I/O module in the
+[King](#the-serf-and-the-king). Therefore there is no need for Ames to do
+something like nack a packet whose checksum does not match the `+mug` of the
```
https://github.com/urbit/docs/pull/957/commits/b9353544c7d1840c7ec464e916efb088462c0e12#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR219

syzygyzer commented on tutorials/arvo/ames.md:
```diff
@@ -112,40 +115,299 @@ ducts as unique keys.  Stop tracking a peer by sending Ames a
+something like nack a packet whose checksum does not match the `+mug` of the
+body.
+
+A nack will be accompanied by a naxplanation, which is a third type of message
```
https://github.com/urbit/docs/pull/957/commits/b9353544c7d1840c7ec464e916efb088462c0e12#diff-b229785bff07fd288a75269882aa1977e802cea1e525ebcf42f4651b428d67daR225

syzygyzer commented:
@belisarius222 Alright I'm finished with this round of revisions. I was trying to handle the review with emacs, but I seemed to have messed something up and switched to just using the GH website partway through, so it got a little fucky, but for comments that were easy to resolve I just closed them and for ones I want you to look at I linked. Hopefully that does not end up being a PITA.

belisarius222 commented:
No, they're not the same; the daemon is the I/O process and the worker is the nock interpreter process.  The daemon writes events to disk, and the worker writes snapshots to disk.

—
~rovnys-ricfer
https://urbit.org

On Wed, Nov 18, 2020 at 5:05 PM, poprox < notifications@github.com > wrote:

> 
> 
> 
> *@syzygyzer* commented on this pull request.
> 
> 
> 
> In tutorials/ arvo/ ames. md (
> https://github.com/urbit/docs/pull/957#discussion_r526453584 ) :
> 
> > +
> +Once `~worwel-sipnum` processes the packet, it will know the IP and
> port of
> +`~bacbel-tagfeb` since `~zod` included it when it forwarded the
> packet and so
> +will send an ack packet directly to `~bacbel-tagfeb`.
> Communication between the
> +two ships will now be direct until one of them
> changes their IP address, port, or networking keys.
> +
> +
> +## The Serf and
> the King
> +
> +Urbit's functionality is split between the two binaries
> `urbit-worker` (sometimes
> +called the Serf) and `urbit-king` (sometimes
> called the King). This division of
> +labor is currently not well-documented
> outside of the [Vere documents](@/docs/tutorials/vere/_index.md), but we
> summarize it here.
> +
> +In short, the Serf is the Nock runtime and so keeps
> track of the current state
> +of Arvo as a Nock noun and updates the state
> by `%poke`ing it with nouns, and
> +then informs the King of the new state.
> The King manages snapshots of the Arvo
> 
> 
> I'm a bit confused - I thought the daemon and the worker were the same
> thing? Maybe I should be more clear of what is meant by manage - does the
> King save the snapshots to disk since it handles I/O? Or does the daemon
> (or worker? whichever you mean) actually also talk to Unix in order to
> save snapshots?
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub (
> https://github.com/urbit/docs/pull/957#discussion_r526453584 ) , or unsubscribe
> (
> https://github.com/notifications/unsubscribe-auth/AAGVR5LVPG5SKUTP3YTJNTDSQRAJJANCNFSM4SGSUOJQ
> ).
> 
> 
>

syzygyzer commented:
What did you mean by "I think you mean `urbit-daemon`, not `urbit-king`" then? If the daemon is I/O process (i.e. the King), then what is the `urbit-king` binary if it is not the daemon?

syzygyzer commented:
Ok I'm not sure how this edit ended up in this commit... I think I must have done something bad with this repo, or my emacs, since this was supposed to just be me changing the names around for the binaries again.
